### ISSUE TITLE
Allow `ns` to have run-time defined namespaces

### ISSUE THREAD

**Issue by ValarDragon** [arkworks-rs/snark#337]
## Summary

Namespace names at the moment must be `&'static str`. There are use cases in which you would want runtime defined namespaces, e.g. for generating many circuits of different sizes, or prefixing what step of a computation you are in.

I think we should try to generalize this to `&str`. I'm not really clear why its limited to `&'static str` at the moment, is this imposed by `tokio-rs/tracing`'s info_span! macro? 

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by Pratyush** [comment]
Yes, the limitation is due to tracing. There might be a way to work around it though
